lib/ampProtection.js

```diff
@@ -0,0 +1,70 @@
+/** @module tds */
+
+const {
+    generateDNRRule
+} = require('./utils')
+
+const AMP_PROTECTION_PRIORITY = 40000
+
+/**
+ * @typedef generateAmpProtectionRulesResult
+ * @property {import('./utils').DNRRule} rule
+ * @property {object} matchDetails
+ */
+
+/**
+ * Generator to produce the declarativeNetRequest rules and corresponding match
+ * details for the given trackerAllowlist configuration.
+ * @param {object} extensionConfiguration
+ *   The extension configuration.
+ * @param {function} isRegexSupported
+ *   A function compatible with chrome.declarativeNetRequest.isRegexSupported.
+ *   See https://developer.chrome.com/docs/extensions/reference/declarativeNetRequest/#method-isRegexSupported
+ * @return {AsyncGenerator<generateAmpProtectionRulesResult>}
+ */
+async function* generateAmpProtectionRules ({ features: { ampLinks } }, isRegexSupported) {
+    if (!ampLinks ||
+        ampLinks.state !== 'enabled' ||
+        !ampLinks.settings ||
+        !ampLinks.settings.linkFormats ||
+        ampLinks.settings.linkFormats.length === 0) {
+        return
+    }
+
+    const { settings: { linkFormats: ampLinkRegexps } } = ampLinks
+    const excludedRequestDomains =
+        (ampLinks?.exceptions || []).map(({ domain }) => domain)
```

Comment:
How do we handle `unprotectedTemporary`, `denylisted` domains, and I thought we had exceptions for `initiator` domains in the TD too? I don't see any of that here.

Reply:
Well spotted, I forgot to add the initiator domain exclusion. Added that now.

As for the unprotectedTemporary allowlist, that works since the AMP rules have a lower priority and so shouldn't match if the unprotectedTemporary allowing rules match instead. The denylisted domains are added as a domain exception for the unprotectedTemprary allowing rules.

Reply:
Can you add a test for the amp priority being lower than the allowlist priority?

Reply:
I've added a check that `AMP_PROTECTION_PRIORITY` === `TRACKING_PARAM_PRIORITY` and there's already checks that `TRACKING_PARAM_PRIORITY` is higher/lower than the right things. I could add this check explicitly, but for consistency I'd be adding I think 7 checks for both GPC and AMP which seems kind of pointless. WDYT?